workspaces/optic-engine/src/queries/shape.rs

```diff
@@ -438,6 +441,96 @@ impl<'a> ShapeQueries<'a> {
     let command = ShapeCommand::remove_field(field_node.field_id.clone());
     Some(std::iter::once(command))
   }
+
+  pub fn edit_shape_commands(
+    &'a self,
+    shape_trail: &ShapeTrail,
+    required_kinds: impl IntoIterator<Item = &'a ShapeKind>,
+    id_generator: &mut impl SpecIdGenerator,
+  ) -> Option<impl Iterator<Item = ShapeCommand>> {
+    let current_trail_choices = self.list_trail_choices(shape_trail);
+
+    let current_kinds: BTreeSet<_> = current_trail_choices
+      .iter()
+      .map(|choice| &choice.core_shape_kind)
+      .cloned()
+      .collect();
+
+    let togglable_kinds = if shape_trail.is_field() {
+      BTreeSet::from_iter(vec![ShapeKind::OptionalKind, ShapeKind::NullableKind])
+    } else {
+      BTreeSet::new() // only allow edits to fields for now
+    };
+
+    let primitive_choice = current_trail_choices
+      .into_iter()
+      .filter(|choice| !togglable_kinds.contains(&choice.core_shape_kind))
+      .next()?;
+
+    let subject_shape_id = if shape_trail.is_field() {
+      primitive_choice.parent_trail.path.iter().rev().find_map(
+        |path_component| match path_component {
+          ShapeTrailPathComponent::OptionalItemTrail { inner_shape_id, .. } => Some(inner_shape_id),
+          ShapeTrailPathComponent::NullableItemTrail { inner_shape_id, .. } => Some(inner_shape_id),
+          ShapeTrailPathComponent::ObjectFieldTrail { field_shape_id, .. } => Some(field_shape_id),
+          _ => None,
+        },
+      )
+    } else {
+      None
+    }?; // nothing to update if there isn't a subject
+
+    let required_kinds: BTreeSet<_> = required_kinds
+      .into_iter()
+      .filter(|kind| togglable_kinds.contains(kind))
+      .cloned()
+      .collect();
+
+    let mut new_shape_prototypes = vec![];
+    let mut root_shape_id = subject_shape_id.clone();
```

Comment:
When we implemented the trail learners and command generation from that, we figured out how useful it can be (especially as we add more complex editing later on) to generate prototypes of shapes first, rather than commands straight away. Most significantly, it allows new id generation to happen and references to those to be made ahead of time, so we can make sure all hierarchy is setup correctly, making the commands easier to write sequentially.

I actually hope to find a generic type of Shape prototype that can be used by all the shape command generators, so we only have to implement the generation of commands once. That's something to look at later, though.